internal/mkcgo/parse.go

```diff
@@ -0,0 +1,346 @@
+package mkcgo
+
+import (
+	"bufio"
+	"cmp"
+	"errors"
+	"os"
+	"slices"
+	"strings"
+)
+
+type fnAttributes struct {
+	tag        string
+	variadic   bool
+	importName string
+}
+
+type attribute struct {
+	name         string
+	description  string
+	hasParameter bool
+	handle       func(string, *fnAttributes)
+}
+
+var attributes = [...]attribute{
+	{
+		name:         "tag",
+		description:  "the function will be loaded together with other functions with the same tag.",
+		hasParameter: true,
+		handle: func(s string, opts *fnAttributes) {
+			opts.tag = s
+		},
+	},
+	{
+		name:         "variadic",
+		description:  "the function has variadic arguments, and its name is a custom wrapper for the actual C name, defined in this attribute.",
+		hasParameter: true,
+		handle: func(s string, opts *fnAttributes) {
+			opts.variadic = true
+			opts.importName = s
+		},
+	},
+}
+
+// Parse parses files listed in fs and extracts all syscall
+// functions listed in sys comments. It returns source files
+// and functions collection *Source if successful.
+func Parse(fs ...string) (*Source, error) {
+	src := &Source{
+		Files: fs,
+	}
+	for _, file := range fs {
+		if err := src.parseFile(file); err != nil {
+			return nil, err
+		}
+	}
+	slices.SortFunc(src.Funcs, func(fi, fj *Func) int {
+		return cmp.Compare(fi.CName, fj.CName)
+	})
+	return src, nil
+}
+
+// parseFile parses file name and extracts all symbols.
+func (src *Source) parseFile(name string) error {
+	file, err := os.Open(name)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	s := bufio.NewScanner(file)
+	var inBlockComment, inEnum bool
```

Comment:
Scanning lines with a state machine (vs. e.g. scanning tokens with recursive descent) seems a little complicated. Specifically, passing a reference to the bool with `processComments(line, &inBlockComment)` is starting to make me wonder.

But... I'm not sure that it's really complicated enough to justify changing the approach. 😄 This isn't a general purpose tool, so it's not as if we need to be concerned about adding more complexity to it. If we e.g. hit an edge case, we can simply not do that.